release/cibuild.py

```diff
@@ -10,163 +10,254 @@
+        should_build_pyinstaller = False,
+
+        has_aws_creds = False,
+        has_twine_creds = False,
+
+        docker_username = "",
+        docker_password = "",
+    ):
+        self.system = system
+        self.root_dir = root_dir
+
+        self.travis_tag = travis_tag
+        self.travis_branch = travis_branch
+        self.travis_pull_request = travis_pull_request
+
+        self.should_build_wheel = should_build_wheel
+        self.should_build_docker = should_build_docker
+        self.should_build_pyinstaller = should_build_pyinstaller
+
+        self.appveyor_repo_tag_name = appveyor_repo_tag_name
+        self.appveyor_repo_branch = appveyor_repo_branch
+        self.appveyor_pull_request_number = appveyor_pull_request_number
+
+        self.has_aws_creds = has_aws_creds
+        self.has_twine_creds = has_twine_creds
+        self.docker_username = docker_username
+        self.docker_password = docker_password
+
+    @classmethod
+    def from_env(klass):
+        return klass(
+            system = platform.system(),
+            root_dir = os.path.normpath(os.path.join(os.path.dirname(__file__), "..")),
+
+            travis_tag = os.environ.get("TRAVIS_TAG", ""),
+            travis_branch = os.environ.get("TRAVIS_BRANCH", ""),
+            travis_pull_request = os.environ.get("TRAVIS_PULL_REQUEST"),
+
+            appveyor_repo_tag_name = os.environ.get("APPVEYOR_REPO_TAG_NAME", ""),
+            appveyor_repo_branch = os.environ.get("APPVEYOR_REPO_BRANCH", ""),
+            appveyor_pull_request_number = os.environ.get("APPVEYOR_PULL_REQUEST_NUMBER"),
+
+            should_build_wheel = "WHEEL" in os.environ,
+            should_build_pyinstaller = "PYINSTALLER" in os.environ,
+            should_build_docker = "DOCKER" in os.environ,
+
+            has_aws_creds = "AWS_ACCESS_KEY_ID" in os.environ,
+            has_twine_creds= (
+                "TWINE_USERNAME" in os.environ and
+                "TWINE_PASSWORD" in os.environ
+            ),
+
+            docker_username = os.environ.get("DOCKER_USERNAME"),
+            docker_password = os.environ.get("DOCKER_PASSWORD"),
+        )
+
+    @property
+    def has_docker_creds(self) -> bool:
+        return self.docker_username and self.docker_password
+
+    @property
+    def is_pull_request(self) -> bool:
+        if self.appveyor_pull_request_number:
+            return True
+        if self.travis_pull_request and self.travis_pull_request != "false":
+            return True
+        return False
+
+    @property
+    def tag(self):
+        return self.travis_tag or self.appveyor_repo_tag_name
+
+    @property
+    def branch(self):
+        return self.travis_branch or self.appveyor_repo_branch
+
+    @property
+    def version(self):
+        name = self.tag or self.branch
+        if not name:
+            raise BuildError("We're on neither a tag nor a branch - could not establish version")
+        return re.sub('^v', "", name)
```

Comment:
I thought about an edge case when I implemented this regex a few days ago, maybe you find a nifty solution:
* branches that start with a `v` but are not a version number.

Is this an issue we need to catch?

Reply:
Yes. In the next revision, I will verify that tags and branches match the specification in PEP 440 (https://www.python.org/dev/peps/pep-0440/). This will let us distinguish between real releases, dev releases and feature branches based on name. Then we can constrain and document these names.